index.coffee

```diff
@@ -148,20 +35,171 @@ onProgressHandler = (onProgressPromise, fallbackOnProgress) ->
-	if not shortId?
-		throw new Error('Progress event missing layer id. Progress not correct.')
-	longId = _.find(layerIds, (id) -> _.startsWith(id, shortId))
-	if not longId?
-		throw new Error("Progress error: Unknown layer #{shortId} referenced by docker. Progress not correct.")
-	return longId
+combinedProgressHandler = (renderer, states, index, callback) ->
+	return (evt) ->
+		# update current reporter state
+		states[index].percentage = evt.percentage
+		# update totals
+		percentage = _.sumBy(states, 'percentage') // (states.length || 1)
+		# update event
+		evt.totalProgress = renderer(percentage)
+		evt.percentage = percentage
+		evt.progressIndex = index
+		# call callback with aggregate event
+		callback(evt)
+
+class ProgressTracker
+	constructor: (@coalesceBelow = 0) ->
+		@layers = {}
+
+	addLayer: (id) ->
+		@layers[id] = { progress: null, coalesced: false }
+
+	updateLayer: (id, progress) ->
+		return if not id?
+		@addLayer(id) if not @layers[id]
+		@patchProgressEvent(progress)
+		@layers[id].coalesced = not @layers[id].progress? and progress.total < @coalesceBelow
+		@layers[id].progress = (progress.current / progress.total) || 0 # prevent NaN when .total = 0
+
+	finishLayer: (id) ->
+		@updateLayer(id, { current: 1, total: 1 })
+
+	getProgress: ->
+		layers = _.filter(@layers, coalesced: false)
+		layerCount = layers.length || 1
+		progress = _.sumBy(layers, 'progress')
+		return Math.round(100 * progress / layerCount)
+
+	patchProgressEvent: (progress) ->
+		# some events arrive without .total
+		progress.total ?= progress.current
+		# some events arrive with .current > .total
+		progress.current = Math.min(progress.current, progress.total)
+
+class ProgressReporter
+	constructor: (@renderProgress) -> #
+
+	# Create a stream that transforms `docker.modem.followProgress` onProgress
+	# events to include total progress metrics.
+	pullProgress: (image, onProgress) ->
+		progressRenderer = @renderProgress
+		downloadProgressTracker = new ProgressTracker(100 * 1024) # 100 KB
+		extractionProgressTracker = new ProgressTracker(1024 * 1024) # 1 MB
+		lastPercentage = 0
+		Promise.try ->
+			return (evt) ->
+				try
+					{ id, status } = evt
+
+					if status is 'Pulling fs layer'
+						downloadProgressTracker.addLayer(id)
+						extractionProgressTracker.addLayer(id)
+					else if status is 'Downloading'
+						downloadProgressTracker.updateLayer(id, evt.progressDetail)
+					else if status is 'Extracting'
+						extractionProgressTracker.updateLayer(id, evt.progressDetail)
+					else if status is 'Download complete'
+						downloadProgressTracker.finishLayer(id)
+					else if status is 'Pull complete'
+						extractionProgressTracker.finishLayer(id)
+					else if status is 'Already exists'
+						downloadProgressTracker.finishLayer(id)
+						extractionProgressTracker.finishLayer(id)
+
+					if status.match(/^Status: Image is up to date for /) or status.match(/^Status: Downloaded newer image for /)
```

Comment:
It's probably better to use `/^.../.test(status)` or `status.startsWith('...')` in order to avoid building up the matches array we don't care about and instead just getting a boolean result